sqlmesh/core/model/definition.py

```diff
@@ -1057,6 +1057,12 @@ def validate_definition(self) -> None:
             if count > 1:
                 raise_config_error(f"Found duplicate outer select name '{name}'", self._path)
 
+        if self.depends_on_past and not self.annotated:
+            raise_config_error(
+                "Self-referencing models require inferrable schemas (columns/types). There are three options: Add types to all outer projections, leverage external models (https://sqlmesh.readthedocs.io/en/stable/concepts/models/external_models/), or use the columns property (https://sqlmesh.readthedocs.io/en/stable/concepts/models/overview/#columns).",
```

Comment:
We should align on our terminology:
"columns property" => "`columns` model attribute"